libraries/AP_OSD/AP_OSD.cpp

```diff
@@ -0,0 +1,111 @@
+/*
+ * This file is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * AP_OSD partially based on betaflight and inav osd.c implemention.
+ * clarity.mcm font is taken from inav configurator.
+ * Many thanks to their authors.
+ */
+
+#include "AP_OSD.h"
+#include "AP_OSD_MAX7456.h"
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/Util.h>
+
+#include <utility>
+
+const AP_Param::GroupInfo AP_OSD::var_info[] = {
+
+    // @Param: _TYPE
+    // @DisplayName: OSD type
+    // @Description: OSD type
+    // @Values: 0:None,1:MAX7456
+    // @User: Standard
+    AP_GROUPINFO_FLAGS("_TYPE", 1, AP_OSD, osd_type, 0, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: _UPDATE_FONT
+    // @DisplayName: Update font
+    // @Description: Undate font inside osd chip
+    // @Values: 0:Do not update,1:Update
+    // @User: Standard
+    AP_GROUPINFO("_UPDATE_FONT", 2, AP_OSD, update_font, 1),
+
+    // @Group: 1_
+    // @Path: AP_OSD_Screen.cpp
+    AP_SUBGROUPINFO(screen[0], "1_", 3, AP_OSD, AP_OSD_Screen),
+
+    // @Group: 2_
+    // @Path: AP_OSD_Screen.cpp
+    AP_SUBGROUPINFO(screen[1], "2_", 4, AP_OSD, AP_OSD_Screen),
+
+    // @Group: 2_
+    // @Path: AP_OSD_Screen.cpp
+    AP_SUBGROUPINFO(screen[2], "3_", 5, AP_OSD, AP_OSD_Screen),
+
+    // @Group: 2_
+    // @Path: AP_OSD_Screen.cpp
+    AP_SUBGROUPINFO(screen[3], "4_", 6, AP_OSD, AP_OSD_Screen),
+    
+    AP_GROUPEND
+};
+
+extern const AP_HAL::HAL& hal;
+
+AP_OSD::AP_OSD()
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    // default first screen enabled
+    screen[0].enabled = 1;
```

Comment:
Doesn't this always force the first OSD to be enabled? (which means you consume all the memory regardless of if you want to use the OSD or not).

Reply:
First osd screen is always enabled. But osd itself has _enabled param which will take precidence

Reply:
@WickedShell I don't think so, this constructor is called before the parameters are loaded from eeprom. It just changes the default for the first screen

Reply:
I confused the interactions of screen objects `_TYPE` parameter here, @sh83 is correct and we can ignore my concern.